### ISSUE TITLE
Allow defining post-run hooks on asset jobs

### ISSUE THREAD

**Issue by askvinni** [dagster-io/dagster#10155]
### What's the use case?

As the `define_asset_job()` method doesn't take a `hooks` argument, the only way to execute a post-run hook is by creating a `RunStatusSensor`.

In a production environment however, each run will typically be running in its own separated environment/pod/ECS task, effectively meaning the sensor can't reliably fetch any information from the run that's saved in the file system of the isolated environment, I'm thinking of e.g. uploading the executed SQL after running a dbt op, but I imagine there's many more use cases.

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by askvinni** [comment]
Apologies, just saw there's already an issue for a similar use case under #8577

**Comment by sryza** [comment]
>  I'm thinking of e.g. uploading the executed SQL after running a dbt op

Curious - are you imagining uploading it as Dagster metadata or to some external system?

**Comment by askvinni** [comment]
@sryza the initial use case would be to upload both the run_results.json as well as the compiled SQL to S3, but I could see attaching them to dagster metadata as well

**Comment by sryza** [comment]
A workaround for the timebeing could be to use the `runtime_metadata_fn` argument of `load_assets_from_dbt_project`. It lets you return metadata that gets attached to the asset materialization.  You could also put code inside of it that writes to S3.

**Comment by askvinni** [comment]
Ah, that's an interesting workaround, definitely fit for my immediate use case.

**Comment by pdstrnadJC** [comment]
This would be great to have. If hooks are exposed when using the `@job` decorator then `define_asset_job()` should allow for hooks too. I guess there must be a way to define an asset-based job "manually" using the job decorator, and then you could also specify hooks - but it's nice to have the convenience of letting Dagster create the job for you based on selected assets. 

However, defining the job "manually" wouldn't work in cases where the assets and their names aren't known until runtime of dagit. For example, I'm generating assets based on configuration, and assets names are derived from that configuration, so I can't refer to the assets, or the functions that define them, in a static job definition.

**Comment by pdstrnadJC** [comment]
It seems like sensors would provide a workaround for my scenario. Indeed I have been wondering what the difference really is between sensors and hooks - I guess using a sensor results in more of an async operation where the sensor is waiting to be triggered, whereas the hook is invoked more directly based on the outcome of the job.

**Comment by cdchan** [comment]
It would be convenient if run results were also passed into `runtime_metadata_fn`. Since it's only the manifest data, `runtime_metadata_fn` can't do anything based on how the run actually went

**Comment by martinv13** [comment]
I can confirm that the solution with `runtime_metadata_fn` works (I use it to parse and load `run_results.json` to the database). However it does not seem to be triggered if for some reason there are only failed dbt runs? Could you confirm this behaviour?

Also is there an updated recommended way to trigger a function after a run of a dbt asset job?

Thanks.

**Comment by axellpadilla** [comment]
this issue should be closed https://docs.dagster.io/api/dagster/hooks

**Comment by bjackson-m3** [comment]
@axellpadilla I don't think so - the hooks apply to ops within the job (i.e. individual assets), but not on the job itself.

**Comment by martinv13** [comment]
For me it is no longer an issue because the new syntax for dbt assets definition (which has been there for quite some time now) does exactly what I needed. See the example [here](https://docs.dagster.io/api/libraries/dagster-dbt#dagster_dbt.dbt_assets) which retrieves and processes dbt artifacts.

A "failure" hook behavior can also be achieved easily by adding a try except within the `@dbt_assets`.

So if the use case is mostly about dbt, we should be already covered. Otherwise and more generally I could not say.

**Comment by axellpadilla** [comment]
@bjackson-m3 you are right, It could be beneficial to add the same on jobs too
